core/src/main/java/org/infinispan/expiration/impl/ExpirationManagerImpl.java

```diff
@@ -252,22 +260,36 @@ private void deleteFromStoresAndNotify(K key, V value, Metadata metadata) {
             cacheNotifier.notifyCacheEntryExpired(key, value, metadata, null)));
    }
 
-   protected Long localLastAccess(Object key, Object value, int segment) {
-      InternalCacheEntry ice = dataContainer.running().peek(segment, key);
-      if (ice != null && (value == null || value.equals(ice.getValue())) &&
-            !ice.isExpired(timeService.wallClockTime())) {
-         return ice.getLastUsed();
+   @Override
+   public CompletionStage<Boolean> handlePossibleExpiration(InternalCacheEntry<K, V> ice, int segment, boolean isWrite) {
+      long currentTime = timeService.wallClockTime();
+      if (ice.isExpired(currentTime)) {
+         if (trace) {
+            log.tracef("Retrieved entry for key %s was expired locally, attempting expiration removal", ice.getKey());
+         }
+         CompletableFuture<Boolean> expiredStage = entryExpiredInMemory(ice, currentTime, isWrite);
+         if (trace) {
+            expiredStage = expiredStage.thenApply(expired -> {
+               if (expired == Boolean.FALSE) {
+                  log.tracef("Retrieved entry for key %s was found to not be expired.", ice.getKey());
+               } else {
+                  log.tracef("Retrieved entry for key %s was confirmed to be expired.", ice.getKey());
+               }
+               return expired;
+            });
+         }
+         return expiredStage;
+      } else if (!isWrite && ice.isMaxIdleExpirable()) {
+         return touchEntry(ice, segment);
       }
-      return null;
+      return CompletableFutures.completedFalse();
    }
 
-   @Override
-   public CompletableFuture<Long> retrieveLastAccess(Object key, Object value, int segment) {
-      Long lastAccess = localLastAccess(key, value, segment);
-      if (lastAccess != null) {
-         return CompletableFuture.completedFuture(lastAccess);
-      }
-      return CompletableFutures.completedNull();
+   protected CompletionStage<Boolean> touchEntry(InternalCacheEntry entry, int segment) {
+      TouchCommand touchCommand = cf.running().buildTouchCommand(entry.getKey(), segment);
+      touchCommand.init(componentRegistry, false);
+      CompletableFuture<Boolean> future = (CompletableFuture) touchCommand.invokeAsync();
+      return future.thenApply(touched -> !touched);
```

Comment:
I'd expect `TouchCommand` and a method named `touchEntry` to use the same convention for the return value.
I suggest moving `thenApply(touched -> !touched)` outside the method.

Reply:
I renamed the method.